File: httpx/_breaker.py
import time
import logging

log = logging.getLogger('crawler.circuit_breaker')


class CircuitBreaker:
    """熔断机制：连续失败>=3次->冷却60秒->自动恢复"""

    OPEN = 'OPEN'
    HALF_OPEN = 'HALF_OPEN'
    CLOSED = 'CLOSED'

    def __init__(self, failure_threshold=3, recovery_timeout=60, name='default'):
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.failure_count = 0
        self.state = self.CLOSED
        self.last_failure_time = 0
        self.name = name

    def call(self, func, *args, **kwargs):
        if self.state == self.OPEN:
            if time.time() - self.last_failure_time >= self.recovery_timeout:
                log.info('[CB:%s] HALF_OPEN', self.name)
                self.state = self.HALF_OPEN
            else:
                raise RuntimeError(f'[CB:{self.name}] OPEN, cooling {self.recovery_timeout}s')
        try:
            result = func(*args, **kwargs)
            if self.state == self.HALF_OPEN:
                log.info('[CB:%s] Recovered', self.name)
                self.state = self.CLOSED
                self.failure_count = 0
            return result
        except Exception as e:
            self.failure_count += 1
            self.last_failure_time = time.time()
            log.warning('[CB:%s] Fail %d/%d: %s', self.name,
                        self.failure_count, self.failure_threshold, str(e)[:60])
            if self.failure_count >= self.failure_threshold:
                log.warning('[CB:%s] OPEN', self.name)
                self.state = self.OPEN
            raise

    def reset(self):
        self.failure_count = 0
        self.state = self.CLOSED

File: go-spider/internal/httpx/breaker.go
package httpx

import (
	"log"
	"sync"
	"time"
)

type BreakerState int

const (
	Closed BreakerState = iota
	Open
	HalfOpen
)

func (s BreakerState) String() string {
	switch s {
	case Closed: return "CLOSED"
	case Open: return "OPEN"
	case HalfOpen: return "HALF_OPEN"
	default: return "UNKNOWN"
	}
}

type CircuitBreaker struct {
	mu               sync.Mutex
	name             string
	state            BreakerState
	failureCount     int
	failureThreshold int
	recoveryTimeout  time.Duration
	lastFailureTime  time.Time
}

func NewBreaker(name string, threshold int, recovery time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		name:             name,
		state:            Closed,
		failureThreshold: threshold,
		recoveryTimeout:  recovery,
	}
}

func (cb *CircuitBreaker) Allow() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case Closed:
		return true
	case Open:
		if time.Since(cb.lastFailureTime) >= cb.recoveryTimeout {
			cb.state = HalfOpen
			log.Printf("[breaker:%s] OPEN -> HALF_OPEN", cb.name)
			return true
		}
		return false
	case HalfOpen:
		return true
	}
	return false
}

func (cb *CircuitBreaker) Success() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.failureCount = 0
	if cb.state == HalfOpen {
		cb.state = Closed
		log.Printf("[breaker:%s] HALF_OPEN -> CLOSED", cb.name)
	}
}

func (cb *CircuitBreaker) Failure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.failureCount++
	cb.lastFailureTime = time.Now()
	if cb.failureCount >= cb.failureThreshold && cb.state != Open {
		cb.state = Open
		log.Printf("[breaker:%s] CLOSED -> OPEN (failures=%d)", cb.name, cb.failureCount)
	}
	if cb.state == HalfOpen {
		cb.state = Open
		log.Printf("[breaker:%s] HALF_OPEN -> OPEN", cb.name)
	}
}

func (cb *CircuitBreaker) State() BreakerState {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	return cb.state
}
